src/Combinatorics/EnumerativeCombinatorics/weak_compositions.jl

```diff
@@ -0,0 +1,182 @@
+################################################################################
+#
+#  Constructors and basic functionality
+#
+################################################################################
+
+data(C::WeakComposition) = C.c
+
+@doc raw"""
+    weak_composition(parts::Vector{T}; check::Bool = true) where T <: IntegerUnion
+
+Return the weak composition given by the integer sequence `parts` as an object of type
+`WeakComposition{T}`.
+
+If `check` is `true` (default), it is checked whether the given sequence defines
+a weak composition, that is, whether all elements of `parts` are non-negative.
+
+# Examples
+```jldoctest
+julia> W = weak_composition([6, 0, 2, 3]) # the weak composition 6, 0, 2, 3 of 11
+[6, 0, 2, 3]
+
+julia> W = weak_composition(Int8[6, 0, 2, 3]) # save the elements in 8-bit integers
+Int8[6, 0, 2, 3]
+```
+"""
+function weak_composition(parts::Vector{T}; check::Bool = true) where {T <: IntegerUnion}
+  if check
+    @req all(x -> x >= 0, parts) "The integers must be non-negative"
+  end
+  return WeakComposition{T}(parts)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", C::WeakComposition)
+  print(io, data(C))
+end
+
+################################################################################
+#
+#  Array-like functionality
+#
+################################################################################
+
+function Base.size(C::WeakComposition)
+  return size(data(C))
+end
+
+function Base.length(C::WeakComposition)
+  return length(data(C))
+end
+
+function Base.getindex(C::WeakComposition, i::IntegerUnion)
+  return getindex(data(C), Int(i))
+end
+
+function Base.setindex!(C::WeakComposition, x::IntegerUnion, i::IntegerUnion)
+  return setindex!(data(C), x, Int(i))
+end
+
+function Base.copy(C::WeakComposition)
+  return weak_composition(copy(data(C)))
+end
+
+################################################################################
+#
+#  Generating weak compositions
+#
+################################################################################
+
+@doc raw"""
+    weak_compositions(n::IntegerUnion, k::IntegerUnion)
+
+Return an iterator over all weak compositions of a non-negative integer `n` into
+`k` parts, produced in lexicographically *descending* order.
+Using a smaller integer type for `n` (e.g. `Int8`) may increase performance.
+
+By a weak composition of `n` into `k` parts we mean a sequence of `k` non-negative
+integers whose sum is `n`.
+
+# Examples
+```jldoctest
+julia> W = weak_compositions(3, 2)
+Iterator over the weak compositions of 3 into 2 parts
+
+julia> length(W)
+4
+
+julia> collect(W)
+4-element Vector{Oscar.WeakComposition{Int64}}:
+ [3, 0]
+ [2, 1]
+ [1, 2]
+ [0, 3]
+```
+"""
+function weak_compositions(n::IntegerUnion, k::IntegerUnion)
+  kk = Int(k)
+  return WeakCompositions(n, kk)
+end
+
+# I have no idea what to call these getter functions
+base(W::WeakCompositions) = W.n
+parts(W::WeakCompositions) = W.k
```

Comment:
Since the fields are called `.n` and `.k`, I prefer getter functions. I could of course also give the fields more meaningful names, but then I still would not know what to call them :)